src/cascadia/TerminalSettingsEditor/NullableColorPicker.cpp

```diff
@@ -0,0 +1,233 @@
+        };
+        double r = linearize(color.R);
+        double g = linearize(color.G);
+        double b = linearize(color.B);
+        return 0.2126 * r + 0.7152 * g + 0.0722 * b;
+    }
+
+    SolidColorBrush NullableColorPicker::CalculateBorderBrush(const Windows::UI::Color& color)
+    {
+        // Accessibility requirement is a 3:1 contrast ratio
+        static const auto whiteLuminance = _CalculateLuminance(Colors::White());
+        const auto contrastRatio = (whiteLuminance + 0.05) / (_CalculateLuminance(color) + 0.05);
+        if (contrastRatio > 3.0)
+        {
+            return SolidColorBrush(Colors::White());
+        }
+        else
+        {
+            return SolidColorBrush(Colors::Black());
+        }
+    }
+
+    void NullableColorPicker::ColorChip_Clicked(const IInspectable& sender, const RoutedEventArgs& /*args*/)
+    {
+        const auto& btn = sender.as<Windows::UI::Xaml::Controls::Primitives::ToggleButton>();
+        const auto& colorEntryColor = btn.DataContext().as<Editor::ColorTableEntry>().Color();
+        const Microsoft::Terminal::Core::Color terminalColor{ colorEntryColor.R, colorEntryColor.G, colorEntryColor.B, colorEntryColor.A };
+        CurrentColor(terminalColor);
+        btn.IsChecked(true);
+    }
+
+    void NullableColorPicker::ColorChip_DataContextChanged(const IInspectable& sender, const DataContextChangedEventArgs& args)
+    {
+        if (const auto& toggleBtn = sender.try_as<Controls::Primitives::ToggleButton>())
+        {
+            if (const auto& currentColor = CurrentColor())
+            {
+                const auto& currentColorVal = currentColor.Value();
+                const auto& newChipColor = args.NewValue().as<Editor::ColorTableEntry>().Color();
+                toggleBtn.IsChecked(newChipColor.R == currentColorVal.R && newChipColor.G == currentColorVal.G && newChipColor.B == currentColorVal.B);
+            }
+        }
+    }
+
+    bool NullableColorPicker::IsNull(IReference<Microsoft::Terminal::Core::Color> color)
+    {
+        return color == nullptr;
+    }
+
+    void NullableColorPicker::NullColorButton_Clicked(const IInspectable& /*sender*/, const RoutedEventArgs& /*args*/)
+    {
+        CurrentColor(nullptr);
+    }
+
+    IAsyncAction NullableColorPicker::MoreColors_Clicked(const IInspectable& /*sender*/, const RoutedEventArgs& /*args*/)
+    {
+        co_await ColorPickerDialog().ShowAsync();
+    }
+
+    void NullableColorPicker::ColorPickerDialog_Opened(const IInspectable& /*sender*/, const ContentDialogOpenedEventArgs& /*args*/)
+    {
+        // Initialize color picker with current color
+        if (CurrentColor())
+        {
+            const auto& terminalColor = CurrentColor().Value();
+            const Windows::UI::Color winuiColor{
+                .A = terminalColor.A,
+                .R = terminalColor.R,
+                .G = terminalColor.G,
+                .B = terminalColor.B
+            };
```

Comment:
The types are incompatible?

Reply:
Yeah, normally I'd use `til::color`, but these are really quick/short comparisons/conversions. No use in actually storing it as a til::color since we have to interact with so much XAML. 😕